File: Woven/Models/Media.swift
import Foundation

// MARK: - Media Type

enum MediaType: String, Codable {
    case photo = "photo"
    case video = "video"
}

// MARK: - Media Model

struct Media: Codable, Identifiable {
    let id: UUID
    let vaultId: UUID
    let mediaType: MediaType
    let fileName: String
    let fileSize: Int
    let uploadedById: Int
    let uploadedBy: MediaUser?
    let createdAt: Date
    
    enum CodingKeys: String, CodingKey {
        case id
        case vaultId = "vault_id"
        case mediaType = "media_type"
        case fileName = "file_name"
        case fileSize = "file_size"
        case uploadedById = "uploaded_by_id"
        case uploadedBy = "uploaded_by"
        case createdAt = "created_at"
    }
    
    var isPhoto: Bool {
        mediaType == .photo
    }
    
    var isVideo: Bool {
        mediaType == .video
    }
}

// MARK: - Media User (nested in media responses)

struct MediaUser: Codable {
    let id: Int
    let username: String?
    let email: String?
    let fullName: String?
    
    enum CodingKeys: String, CodingKey {
        case id, username, email
        case fullName = "full_name"
    }
}

// MARK: - Media List Response

struct MediaListResponse: Codable {
    let media: [Media]
    let total: Int
}

// MARK: - Media View URL Response

struct MediaViewUrlResponse: Codable {
    let viewUrl: String
    let expiresIn: Int
    
    enum CodingKeys: String, CodingKey {
        case viewUrl = "view_url"
        case expiresIn = "expires_in"
    }
}

File: backend/app/schemas/media.py
from pydantic import BaseModel
from typing import Optional, Literal
from datetime import datetime
from uuid import UUID

from app.schemas.user import UserResponse


class MediaUploadRequest(BaseModel):
    """Request to get upload URL for media."""
    vault_id: UUID
    file_name: str
    file_size: int  # Size in bytes
    media_type: Literal["photo", "video"]
    encryption_iv: str  # Base64 encoded IV
    encryption_tag: str  # Base64 encoded authentication tag


class MediaUploadResponse(BaseModel):
    """Response with upload URL and media ID."""
    media_id: UUID
    upload_url: str
    expires_in: int  # Seconds until URL expires


class MediaCreate(BaseModel):
    """Request to register uploaded media."""
    vault_id: UUID
    storage_key: str
    file_name: str
    file_size: int
    media_type: Literal["photo", "video"]
    encryption_iv: str
    encryption_tag: str
    thumbnail_key: Optional[str] = None


class MediaResponse(BaseModel):
    """Media metadata in API responses."""
    id: UUID
    vault_id: UUID
    media_type: str
    file_name: str
    file_size: int
    uploaded_by_id: int
    uploaded_by: Optional[UserResponse] = None
    created_at: datetime
    
    class Config:
        from_attributes = True


class MediaViewUrlResponse(BaseModel):
    """Response with temporary view URL for media."""
    view_url: str
    expires_in: int  # Seconds until URL expires


class MediaListResponse(BaseModel):
    """List of media in a vault."""
    media: list[MediaResponse]
    total: int


